community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/storage/paged/PagedIndexInput.java

```diff
@@ -0,0 +1,531 @@
+
+            moveCursorTo( pageId + 1, 0 );
+            do
+            {
+                b4 = cursor.getByte( 0 );
+            }
+            while ( cursor.shouldRetry() );
+        }
+
+        if ( cursor.checkAndClearBoundsFlag() )
+        {
+            throw newEOFException( "Cursor bounds check failed" );
+        }
+
+        return ((b1 & 0xFF) << 24) | ((b2 & 0xFF) << 16) | ((b3 & 0xFF) << 8) | (b4 & 0xFF);
+    }
+
+    @Override
+    public long readLong() throws IOException
+    {
+        long val = readLong( currentPageId, currentPageOffset );
+        incrementPageOffset( 8 );
+        return val;
+    }
+
+    @Override
+    public long readLong( long pos ) throws IOException
+    {
+        return readLong( pageId( pos ), pageOffset( pos ) );
+    }
+
+    private long readLong( long pageId, int pageOffset ) throws IOException
+    {
+        // Cross-page read?
+        if ( pageSize - pageOffset < 8 )
+        {
+            // Delegate to readInt, which delegates to readPageCrossingInt
+            int secondPageOffset = (pageOffset + 4) & moduloPageSize;
+            long secondPage = secondPageOffset < pageOffset ? pageId + 1 : pageId;
+
+            int int1 = readInt( pageId, pageOffset );
+            int int2 = readInt( secondPage, secondPageOffset );
+            return (((long) int1) << 32) | (int2 & 0xFFFFFFFFL);
+        }
+
+        // Regular read
+        moveCursorTo( pageId, pageOffset );
+
+        long val;
+        do
+        {
+            val = cursor.getLong( pageOffset );
+        }
+        while ( cursor.shouldRetry() );
+
+        if ( cursor.checkAndClearBoundsFlag() )
+        {
+            throw newEOFException( "Cursor bounds check failed" );
+        }
+
+        return val;
+    }
+
+    @Override
+    public final void readBytes( byte[] b, int offset, int len ) throws IOException
+    {
+        int bytesRead = 0;
+
+        while ( bytesRead < len )
+        {
+            moveCursorTo( currentPageId, currentPageOffset );
+
+            int toRead = Math.min( len - bytesRead, pageSize - currentPageOffset );
+            do
+            {
+                cursor.setOffset( currentPageOffset );
+                cursor.getBytes( b, offset + bytesRead, toRead );
+            }
+            while ( cursor.shouldRetry() );
+
+            if ( cursor.checkAndClearBoundsFlag() )
+            {
+                throw newEOFException( "Cursor bounds check failed" );
+            }
+
+            incrementPageOffset( toRead );
+            bytesRead += toRead;
+        }
+    }
+
+    @Override
+    public void seek( long pos ) throws IOException
+    {
+        long newPageId = pageId( pos );
+        int newOffset = pageOffset( pos );
+
+        if ( isEOF( newPageId, newOffset ) )
+        {
+            throw newEOFException(
+                    String.format( "seek EOF check failed for { pos: %d, pageId: %d, offsetInPage: %d }", pos,
+                            newPageId, newOffset ) );
```

Comment:
With all these exceptions, try extracting all of the exception construction to a separate method. For instance in this case you could extract a `newSeekEofException( pos, newPageId, newOffset )`. That would reduce the amount of bytecode in the method, and give the JIT inliner more material to work with.